darts/models/components/layer_norm_variants.py

```diff
@@ -54,3 +54,82 @@ def __init__(self, input_size, **kwargs):
 class LayerNorm(nn.LayerNorm):
     def __init__(self, input_size, **kwargs) -> None:
         super().__init__(input_size, **kwargs)
+
+
+class ReversibleInstanceNorm(nn.Module):
+    """Reversible Instance Normalization based on [1]
+
+    References
+    ----------
+    .. [1] Kim et al. "Reversible Instance Normalization for Accurate Time-Series Forecasting against
+            Distribution Shift" International Conference on Learning Representations (2022)
+    """
+
+    def __init__(self, axis, input_dim, eps=1e-5, affine=True):
+        super().__init__()
+        self.axis = axis
+        self.input_dim = input_dim
+        self.eps = eps
+        self.affine = affine
+
+        if self.affine:
+            self.affine_weight = nn.Parameter(torch.ones(self.input_dim))
+            self.affine_bias = nn.Parameter(torch.zeros(self.input_dim))
+
+    def forward(self, x, mode, target_slice=None):
+        if mode == "norm":
+            self._get_statistics(x)
+            x = self._normalize(x)
+        elif mode == "denorm":
+            x = self._denormalize(x, target_slice)
+        else:
+            raise NotImplementedError
+        return x
+
+    def _get_statistics(self, x):
+        self.mean = torch.mean(x, dim=self.axis, keepdim=True)
+        self.std = torch.sqrt(torch.var(x, dim=self.axis, keepdim=True) + self.eps)
+
+    def _normalize(self, x):
+        x = x - self.mean
+        x = x / self.std
+        if self.affine:
+
+            massage_shape = not (
+                (self.axis is not -1) and (self.axis is not x.ndim - 1)
+            )
+
+            # if axis isn't the last dimension, swap it to the last dimension
+            if massage_shape:
+                x = x.swapaxes(-2, self.axis)
```

Comment:
I seem to recall there was an issue with the matrix multiplication not playing nice if the normalization axis wasn't the last dimension. 

I added tests to cover this because it was a problem. The original source code I was basing my implementation off of assumed the time and feature dimensions were swapped. I will double check this, but it should be correct, but there may be a faster implementation.